task 3: typecheck

migeed-z opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61310
* #61265
* #61239
* #61238
* #61188
* #61150
* #61093
* #61012
* **#60805**

Differential Revision: [D29522885](https://our.internmc.facebook.com/intern/diff/D29522885)

migeed-z committed:
fix mypy errors

[ghstack-poisoned]

migeed-z committed:
Update on "fix mypy errors"




[ghstack-poisoned]

migeed-z committed:
Update on "fix mypy errors"




[ghstack-poisoned]

migeed-z committed:
Update on "fix mypy errors"




[ghstack-poisoned]

migeed-z committed:
Update on "fix mypy errors"




[ghstack-poisoned]

migeed-z committed:
Update on "fix mypy errors"




[ghstack-poisoned]

migeed-z committed:
Update on "fix mypy errors"




[ghstack-poisoned]

migeed-z committed:
Update on "fix mypy errors"




[ghstack-poisoned]

migeed-z committed:
Update on "fix mypy errors"




[ghstack-poisoned]

migeed-z committed:
Update on "fix mypy errors"




[ghstack-poisoned]

migeed-z committed:
Update on "fix mypy errors"




[ghstack-poisoned]

migeed-z committed:
Update on "task 3: typecheck"




[ghstack-poisoned]

migeed-z committed:
Update on "task 3: typecheck"




[ghstack-poisoned]

migeed-z committed:
Update on "task 3: typecheck"




[ghstack-poisoned]

migeed-z committed:
Update on "task 3: typecheck"




[ghstack-poisoned]

jamesr66a commented on torch/fx/experimental/graph_gradual_typechecker.py:
```diff
@@ -0,0 +1,114 @@
+                return False
+        return True
+
+    def type_check_node(self, n):
```
To help improve readability, can you add type annotations to the function parameters here and on collect_params, as well as add a short comment describing what is expected for each of the arguments?

migeed-z replied:
collect_params is removed now, clarified what n is in a comment, however if I annotate it with Node, mypy fails with 

`error: Item "Tuple[Any, ...]" of "Union[Tuple[Any, ...], List[Any], Dict[str, Any], slice, Node, str, int, float, bool, dtype, Tensor, device, None]" has no attribute "type"`

is it possible that not all nodes have the type attribute?

jamesr66a commented on torch/fx/experimental/graph_gradual_typechecker.py:
```diff
@@ -0,0 +1,114 @@
+                return n.type
+
+        if n.op == 'call_function':
+            if n.target == torch.add or n.target == operator.add:
```
To make this a bit cleaner, we could define a system where we can register these rules as a separate function. For example:

```
import torch, operator
import torch.fx
from typing import Callable, Dict

_INFERENCE_RULES : Dict[torch.fx.node.Target, Callable] = {}

def register_inference_rule(call_target):
    def register(fn):
        if call_target in _INFERENCE_RULES:
            raise RuntimeError('Inference rule already registered for {call_target}!')
        _INFERENCE_RULES[call_target] = fn
        return fn
    return register


@register_inference_rule(torch.add)
@register_inference_rule(operator.add)
def add_inference_rule(n : torch.fx.Node):
    # Type inference rule for add here


@register_inference_rule(torch.reshape)
def reshape_inference_rule(n : torch.fx.Node):
    # Type inference for reshape here
```

Then here we can replace the `if/else` tree with something like:

```
if n.target in _INFERENCE_REGISTRY:
  return _INFERENCE_REGISTRY[n.target](n)
else:
  raise RuntimeError(f'No inference rule registered for target {n.target}!')
```

jamesr66a commented on torch/fx/experimental/graph_gradual_typechecker.py:
```diff
@@ -0,0 +1,114 @@
+
+                if is_consistent(t1, t2):
+                    # we return the more precise type
+                    if is_more_precise(t1, t2):
```
So is this what you were referring to in the meeting about putting the more precise type? I was thinking of that in terms of propagating back, but I think propagating forward, you would actually need to encode semantics for broadcasting. For example, you could have shape `1` on one side but `Dyn` on the other, but depending on the situation the `1` could be broadcasted out to the dynamic size

migeed-z replied:
broadcasting is done, let me know if the current definitions and tests are okay!

jamesr66a reviewed (commented):
Looks good! had a few questions/suggestions

migeed-z committed:
Update on "task 3: typecheck"




[ghstack-poisoned]

migeed-z committed:
Update on "task 3: typecheck"




[ghstack-poisoned]

migeed-z committed:
Update on "task 3: typecheck"




[ghstack-poisoned]

migeed-z committed:
Update on "task 3: typecheck"




[ghstack-poisoned]

jamesr66a reviewed (approved):
Looks good!

migeed-z committed:
Update on "task 3: typecheck"




[ghstack-poisoned]

migeed-z committed:
Update on "task 3: typecheck"




[ghstack-poisoned]

migeed-z committed:
Update on "task 3: typecheck"




[ghstack-poisoned]

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29522885).

migeed-z committed:
Update on "task 3: typecheck"


Differential Revision: [D29522885](https://our.internmc.facebook.com/intern/diff/D29522885)

[ghstack-poisoned]

migeed-z committed:
Update on "task 3: typecheck"


Differential Revision: [D29522885](https://our.internmc.facebook.com/intern/diff/D29522885)

[ghstack-poisoned]
